imagetest/test_suites/packagevalidation/package_test.go

```diff
@@ -110,56 +113,113 @@ func TestGuestPackages(t *testing.T) {
-	}
-	if strings.Contains(image, "ubuntu") {
-		requiredPkgs = removeFromArray(requiredPkgs, "gce-disk-expand")
-	}
-
-	installedPkgs, err := listPkgs()
+	pkgs = []*osPackage{
+		&osPackage{
+			name: "google-guest-agent",
+		},
+		&osPackage{
+			name: "google-osconfig-agent",
+		},
+		&osPackage{
+			name:       "google-compute-engine",
+			imagesSkip: []string{"sles", "suse"},
+		},
+		&osPackage{
+			name:   "google-guest-configs",
+			images: []string{"sles", "suse"},
+		},
+		&osPackage{
+			name:   "google-guest-oslogin",
+			images: []string{"sles", "suse"},
+		},
+		&osPackage{
+			name:       "gce-disk-expand",
+			imagesSkip: []string{"sles", "suse", "ubuntu"},
+		},
+		&osPackage{
+			name:         "google-cloud-cli",
+			alternatives: []string{"google-cloud-sdk"},
+		},
+		&osPackage{
+			name:       "google-compute-engine-oslogin",
+			imagesSkip: []string{"sles", "suse"},
+		},
+		&osPackage{
+			name:   "epel-release",
+			images: []string{"centos-7", "rhel-7"},
+		},
+		&osPackage{
+			name:   "haveged",
+			images: []string{"debian"},
+		},
+		&osPackage{
+			name:   "net-tools",
+			images: []string{"debian"},
+		},
+		&osPackage{
+			name:   "google-cloud-packages-archive-keyring",
+			images: []string{"debian"},
+		},
+		&osPackage{
+			name:   "isc-dhcp-client",
+			images: []string{"debian"},
+		},
+		&osPackage{
+			name:                 "cloud-initramfs-growroot",
+			shouldNotBeInstalled: true,
+			images:               []string{"debian"},
+		},
+	}
+
+	installedList, err := listPkgs()
 	if err != nil {
 		t.Errorf("Failed to execute list packages command: %v", err)
 		return
 	}
 
-	missingPkgs := findUniq(requiredPkgs, installedPkgs)
-	for _, pkg := range missingPkgs {
-		t.Errorf("Required package not installed: %s\n", pkg)
+	installedMap := make(map[string]bool)
+	for _, curr := range installedList {
+		installedMap[curr] = true
 	}
 
-	shouldnotPkgs := findUniq(excludePkgs, installedPkgs) // find whats installed
-	shouldnotPkgs = findUniq(excludePkgs, shouldnotPkgs)  // it shouldn't be
-	for _, pkg := range shouldnotPkgs {
-		t.Errorf("Package installed but should not be: %s\n", pkg)
-	}
+	for _, curr := range pkgs {
+		skipPackage := false
+		for _, skipExpression := range curr.imagesSkip {
+			if strings.Contains(image, skipExpression) {
+				skipPackage = true
+				break
+			}
+		}
 
+		imageMatched := len(curr.images) == 0
+		for _, matchExpression := range curr.images {
+			if strings.Contains(image, matchExpression) {
+				imageMatched = true
+				break
+			}
+		}
+
+		if skipPackage || !imageMatched {
+			continue
+		}
+
+		packageInstalled := false
+		packageNames := []string{curr.name}
+		packageNames = append(packageNames, curr.alternatives...)
+
+		for _, currPackage := range packageNames {
+			if _, found := installedMap[currPackage]; found {
+				packageInstalled = true
+				break
+			}
+		}
+
+		if curr.shouldNotBeInstalled && packageInstalled {
```

Comment:
There's a lot of logic here for checking the same two variables, can we just do something like
```
if !curr.shouldNotBeinstalled != packageInstalled {
   t.Errorf("package %s has wrong installation state, got %t want %t [..]
```

Reply:
Yep, we can do that.

Reply:
done. thanks.